shared/src/sources/balancer/swap/stable_math.rs

```diff
@@ -0,0 +1,495 @@
+        for balance in &balances[1..] {
+            // P_D = Math.div(Math.mul(Math.mul(P_D, balances[j]), numTokens), invariant, roundUp);
+            p_d = rounded_div(
+                p_d.bmul(balance.as_uint256())?.bmul(num_tokens)?,
+                invariant,
+                round_up,
+            )?
+        }
+        let prev_invariant = invariant;
+
+        invariant = rounded_div(
+            // invariant = Math.div(
+            //     Math.mul(Math.mul(numTokens, invariant), invariant).add(
+            //         Math.div(Math.mul(Math.mul(ampTimesTotal, sum), P_D), _AMP_PRECISION, roundUp)
+            //     ),
+            num_tokens
+                .bmul(invariant)?
+                .bmul(invariant)?
+                .badd(rounded_div(
+                    amp_times_total.bmul(sum)?.bmul(p_d)?,
+                    *AMP_PRECISION,
+                    round_up,
+                )?)?,
+            // Math.mul(numTokens + 1, invariant).add(
+            //     // No need to use checked arithmetic for the amp precision, the amp is guaranteed to be at least 1
+            //     Math.div(Math.mul(ampTimesTotal - _AMP_PRECISION, P_D), _AMP_PRECISION, !roundUp)
+            // ),
+            (num_tokens.badd(1.into())?)
+                .bmul(invariant)?
+                .badd(rounded_div(
+                    (amp_times_total.bsub(*AMP_PRECISION)?).bmul(p_d)?,
+                    *AMP_PRECISION,
+                    !round_up,
+                )?)?,
+            round_up,
+        )?;
+
+        match convergence_criteria(invariant, prev_invariant) {
+            None => continue,
+            Some(invariant) => return Ok(invariant),
+        }
+    }
+
+    Err(Error::StableInvariantDidntConverge)
+}
+
+fn calc_out_given_in(
+    amplification_parameter: U256,
+    balances: &mut [Bfp],
+    token_index_in: usize,
+    token_index_out: usize,
+    token_amount_in: Bfp,
+) -> Result<Bfp, Error> {
+    let invariant = calculate_invariant(amplification_parameter, balances, true)?;
+    // Ensure no index error at token_index_in
+    let balance_token_in = balances.get(token_index_in).ok_or(Error::InvalidToken)?;
+    balances[token_index_in] = balance_token_in.add(token_amount_in)?;
+
+    let final_balance_out = get_token_balance_given_invariant_and_all_other_balances(
+        amplification_parameter,
+        balances,
+        invariant,
+        token_index_out,
+    )?;
+    // No need to use checked arithmetic since `tokenAmountIn` was actually added to the same balance right before
+    // calling `_getTokenBalanceGivenInvariantAndAllOtherBalances` which doesn't alter the balances array.
+    balances[token_index_in] = balances[token_index_in]
+        .sub(token_amount_in)
+        .expect("will not underflow");
+
+    balances
+        .get(token_index_out)
+        .ok_or(Error::InvalidToken)?
+        .sub(final_balance_out)?
+        .sub(Bfp::from_wei(1.into()))
```

Comment:
Potential problem here is that you might update balances[token_index_in] first and then error afterwards. The function would return an error but the input has changed. Whether this is a problem depends on how it is called. Either way should be documented.

Reply:
I am not exactly sure how I am supposed to go about doing this... I would actually prefer to simply assert in the beginning that `balances.len()` is greater than the two indices being passed in here. is it even possible to mutate the array's length? Ideally I could use getters everywhere, but this make the self addition and subtraction very challenging (to the extent that I couldn't figure out how). 

Also, not sure I understand the potential problem you have pointed out here. Note that we access `token_index_in` first and then `token_index_out` later.

Reply:
What I mean is that in line 98 you modify a value in `balances`. But then line 104 could return an error. In that case the function has returned an error but it has still modified the input. That might be unexpected for someone calling it because they think no modifications take place if there is an error.

>  would actually prefer to simply assert in the beginning

Sounds fine.